File: examples/cs/gmconfig/gmconfig.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/*! \file gmconfig.cs
*
*  A C# example program demonstrating the use of a ConfigFile.
*  (Loading form a XML file, Creating a Connection, Subscribing & Publishing)
*/


namespace gmconfig
{
	using GMSEC.API;
	using GMSEC.API.UTIL;
	using GMSEC.API.Example.Common;
	using System;
    using System.Collections.Generic;

	class gmconfig : GmsecExample
	{
		private ConfigFile cfgFile;
		private Connection conn;
	    private List<SubscriptionInfo> info = new List<SubscriptionInfo>();

		public gmconfig()
		{
			Config tmpConfig = new Config();
			Initialize(tmpConfig);
		}


		private bool printUsage(int numberArgs)
		{
			if (numberArgs < 1)
			{
				Console.Out.WriteLine("\nUsage: gmconfig_cs <filename>\n");
				return true;
			}
			return false;
		}


		private void Run(string[] args)
		{
			if (printUsage(args.Length))
			{
				return;
			}

			try
			{
				// Create the ConfigFile
				cfgFile = new ConfigFile();

				// Load Config File
				cfgFile.Load(args[0]);

				// Lookup Configs from config file
				Config config1 = cfgFile.LookupConfig("config1");
				Config config2 = cfgFile.LookupConfig("config2");

				// Display XML representation of the Configs
				Log.Info("Config 1:\n" + config1.ToXML());
				Log.Info("Config 2:\n" + config2.ToXML());

				// Create connection and connect
				conn = Connection.Create(config2);
				conn.Connect();

				// Lookup subscription patterns from config file and subscribe 
				string pattern1 = cfgFile.LookupSubscription("events");
				string pattern2 = cfgFile.LookupSubscription("custom1");

				Log.Info("Subscribing to " + pattern1);
                info.Add(conn.Subscribe(pattern1));

				Log.Info("Subscribing to " + pattern2);
                info.Add(conn.Subscribe(pattern2));

				// Lookup Message from config file
				Message msg = cfgFile.LookupMessage("msg1");

				// Display XML representation of the message
				Log.Info("Publishing Message:\n" + msg.ToXML());

				// Publish the message
				conn.Publish(msg);
			}
			catch (GMSEC_Exception e)
			{
				Log.Error("Exception: " + e.ToString());
			}
			finally
			{
				if (conn != null)
				{
                    for (int i = info.Count - 1; i >= 0; i--)
                    {
                        Log.Info("Unsubscribing to " + info[i].GetSubject());
                        var temp = info[i];
                        conn.Unsubscribe(ref temp);
                        info.RemoveAt(i);
                    }
                    
                    // Disconnect connection
					conn.Disconnect();

					// Destroy connection
					Connection.Destroy(ref conn);
				}
			}
        }


		static void Main(string[] args)
		{
			new gmconfig().Run(args);
		}
	}
}

File: examples/cpp/gmconfig/gmconfig.cpp
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmconfig.cpp
 * 
 * A C++ example demonstration of loading a configuration from a standard
 * xml file.  The configuration file used for the example code is "gmconfig.xml".
 * 
 */

#include "../example.h"

#include <string>


using namespace gmsec::api;


class gmconfig
{
public:
	gmconfig(const char* filename);
	~gmconfig();
	bool run();

private:
	std::string cfgFilename;
	ConfigFile  cfgFile;
};


gmconfig::gmconfig(const char* filename)
	: cfgFilename(filename),
	  cfgFile()
{
	Config config;
	example::initialize(config);
}


gmconfig::~gmconfig()
{
}


bool gmconfig::run()
{
	bool success = true;

	try
	{
		// Attempt to load configuration file
		cfgFile.load(cfgFilename.c_str());

		// Attempt to retrieve configurations from the configuration file
		Config c1 = cfgFile.lookupConfig("config1");
		Config c2 = cfgFile.lookupConfig("config2");

		// Display log of XML representation of configuration objects
		GMSEC_INFO << "Config 1:\n" << c1.toXML();
		GMSEC_INFO << "Config 2:\n" << c2.toXML();

		// Attempt to create a Connection object using a configuration object
		Connection* connection = Connection::create(c2);

		// Destroy the connection object
		Connection::destroy(connection);

		// Attempt to lookup subscription topic from configuration file
		const char* pattern = cfgFile.lookupSubscription("events");

		if (!pattern)
		{
			throw Exception(CONFIGFILE_ERROR,
			                OTHER_ERROR_CODE,
			                "Failed to acquire 'events' subscription topic");
		}
		GMSEC_INFO << "Subscription topic: " << pattern;

		// Attempt to lookup message from configuration file
		Message message = cfgFile.lookupMessage("msg1");

		// Display XML representation of the message
		GMSEC_INFO << "Message:\n" << message.toXML();

		// Obtain Configuration File Iterator
		ConfigFileIterator& iter = cfgFile.getIterator();

		// Acquire and display all configuration entries
		while (iter.hasNextConfig())
		{
			const ConfigFile::ConfigEntry& entry = iter.nextConfig();

			GMSEC_INFO << "\nConfig Name: " << entry.getName()
			           << "\nConfig     :\n" << entry.getConfig().toXML();
		}

		// Acquire and display all message entries
		while (iter.hasNextMessage())
		{
			const ConfigFile::MessageEntry& entry = iter.nextMessage();

			GMSEC_INFO << "\nMessage Name: " << entry.getName()
			           << "\nMessage     :\n" << entry.getMessage().toXML();
		}

		// Acquire and display all subscription entries
		while (iter.hasNextSubscription())
		{
			const ConfigFile::SubscriptionEntry& entry = iter.nextSubscription();

			GMSEC_INFO << "\nSubscription Name : " << entry.getName()
			           << "\nSubscription Topic: " << entry.getSubject();
		}

		// Acquire and display all custom XML entries
		while (iter.hasNextCustomElement())
		{
			const char* element = iter.nextCustomElement();

			GMSEC_INFO << "\nCustom XML Element:\n" << element;
		}
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


int main(int argc, char* argv[])
{
	if (argc <= 1)
	{
		std::cout << "usage: gmconfig <filename>" << std::endl;
		return -1;
	}

	return (gmconfig(argv[1]).run() ? 0 : -1);
}
